File: app/src/main/java/pl/piotrskiba/dailywallpaper/utils/NetworkUtils.java
package pl.piotrskiba.dailywallpaper.utils;

import android.content.Context;
import android.content.SharedPreferences;
import android.net.Uri;
import android.preference.PreferenceManager;

import java.io.IOException;
import java.io.InputStream;
import java.net.HttpURLConnection;
import java.net.MalformedURLException;
import java.net.URL;
import java.util.Scanner;

import pl.piotrskiba.dailywallpaper.BuildConfig;
import pl.piotrskiba.dailywallpaper.R;

public class NetworkUtils {

    private final static String API_KEY = BuildConfig.API_KEY;

    private final static String BASE_API_URL = "https://pixabay.com/api/";

    private final static String PARAM_API_KEY = "key";
    private final static String PARAM_IMAGE_TYPE = "image_type";
    private final static String PARAM_ORIENTATION = "orientation";
    private final static String PARAM_CATEGORY = "category";
    private final static String PARAM_EDITORS_CHOICE = "editors_choice";
    private final static String PARAM_SAFE_SEARCH = "safesearch";
    private final static String PARAM_ORDER = "order";
    private final static String PARAM_PAGE = "page";
    private final static String PARAM_PER_PAGE = "per_page";

    private final static String VALUE_ORIENTATION = "vertical";
    private final static String VALUE_EDITORS_CHOICE = "true";
    private final static String VALUE_PER_PAGE = "200";

    public static URL buildUrl(String category, boolean safesearch){
        Uri.Builder uriBuilder = Uri.parse(BASE_API_URL).buildUpon()
                .appendQueryParameter(PARAM_API_KEY, API_KEY)
                .appendQueryParameter(PARAM_ORIENTATION, VALUE_ORIENTATION)
                .appendQueryParameter(PARAM_EDITORS_CHOICE, VALUE_EDITORS_CHOICE)
                .appendQueryParameter(PARAM_SAFE_SEARCH, String.valueOf(safesearch))
                .appendQueryParameter(PARAM_PER_PAGE, VALUE_PER_PAGE);

        if(category != null){
            uriBuilder.appendQueryParameter(PARAM_CATEGORY, category);
        }

        Uri uri = uriBuilder.build();

        URL builtUrl = null;

        try {
            builtUrl = new URL(uri.toString());
        } catch (MalformedURLException e) {
            e.printStackTrace();
        }


        return builtUrl;
    }

    public static String getHttpResponse(URL url) throws IOException {
        HttpURLConnection connection = (HttpURLConnection) url.openConnection();

        try {
            InputStream icStream = connection.getInputStream();

            Scanner scanner = new Scanner(icStream);
            scanner.useDelimiter("\\A");

            if (scanner.hasNext()) {
                return scanner.next();
            } else {
                return null;
            }
        }
        finally{
            connection.disconnect();
        }
    }

}

File: app/src/main/java/pl/piotrskiba/dailywallpaper/utils/NetworkUtils.kt
package pl.piotrskiba.dailywallpaper.utils

import android.content.Context
import android.graphics.Bitmap
import android.graphics.BitmapFactory
import androidx.preference.PreferenceManager
import pl.piotrskiba.dailywallpaper.Constants
import pl.piotrskiba.dailywallpaper.R
import pl.piotrskiba.dailywallpaper.interfaces.ImageListLoadedListener
import pl.piotrskiba.dailywallpaper.models.ImageList
import pl.piotrskiba.dailywallpaper.network.PixabayAPIClient
import pl.piotrskiba.dailywallpaper.network.PixabayAPIInterface
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.io.IOException
import java.net.HttpURLConnection
import java.net.URL

object NetworkUtils {

    /*
        get Bitmap from URL
        source: https://stackoverflow.com/questions/8992964/android-load-from-url-to-bitmap
    */
    @JvmStatic
    fun getBitmapFromURL(src: String): Bitmap? {
        return try {
            val url = URL(src)
            val connection = url.openConnection() as HttpURLConnection
            connection.doInput = true
            connection.connect()
            val input = connection.inputStream
            BitmapFactory.decodeStream(input)
        } catch (e: IOException) {
            null
        }
    }

    fun getAllImages(context: Context, listener: ImageListLoadedListener) {
        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
        val safesearch = sharedPreferences.getBoolean(context.getString(R.string.pref_safesearch_key), true)

        val api = PixabayAPIClient.retrofit.create(PixabayAPIInterface::class.java)
        api.loadAllImages(
                Constants.DEFAULT_VALUE_API_KEY,
                Constants.DEFAULT_VALUE_ORIENTATION,
                Constants.DEFAULT_VALUE_EDITORS_CHOICE,
                safesearch
        ).enqueue(object: Callback<ImageList> {
            override fun onFailure(call: Call<ImageList>, t: Throwable) {
                t.printStackTrace()
                listener.onImageListLoadingError()
            }

            override fun onResponse(call: Call<ImageList>, response: Response<ImageList>) {
                response.body()?.let {
                    listener.onImageListLoaded(response.body()!!)
                } ?: kotlin.run {
                    listener.onImageListLoadingError()
                }
            }
        })
    }

    fun getCategoryImages(context: Context, category: Int, listener: ImageListLoadedListener) {
        val allCategories = context.resources.getStringArray(R.array.pref_category_values)
        val categoryName = allCategories[category]

        val sharedPreferences = PreferenceManager.getDefaultSharedPreferences(context)
        val safesearch = sharedPreferences.getBoolean(context.getString(R.string.pref_safesearch_key), true)

        val api = PixabayAPIClient.retrofit.create(PixabayAPIInterface::class.java)
        api.loadCategoryImages(
                Constants.DEFAULT_VALUE_API_KEY,
                Constants.DEFAULT_VALUE_ORIENTATION,
                Constants.DEFAULT_VALUE_EDITORS_CHOICE,
                safesearch,
                categoryName
        ).enqueue(object: Callback<ImageList> {
            override fun onFailure(call: Call<ImageList>, t: Throwable) {
                t.printStackTrace()
                listener.onImageListLoadingError()
            }

            override fun onResponse(call: Call<ImageList>, response: Response<ImageList>) {
                response.body()?.let {
                    listener.onImageListLoaded(response.body()!!)
                } ?: kotlin.run {
                    listener.onImageListLoadingError()
                }
            }
        })
    }
}